src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs

```diff
@@ -0,0 +1,91 @@
+﻿using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+
+namespace Workleap.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+internal sealed class CacheTokenOnStartupBackgroundService : BackgroundService
+{
+    private readonly IHostApplicationLifetime _applicationLifetime;
+    private readonly IClientCredentialsTokenManagementService _tokenManagementService;
+    private readonly IOptions<CacheTokenOnStartupBackgroundServiceOptions> _backgroundServiceOptions;
+    private readonly IOptionsMonitor<ClientCredentialsOptions> _clientCredentialsOptionsMonitor;
+    private readonly TaskCompletionSource<bool> _allTokensCachedSignal;
+    private readonly List<string> _clientNames;
+
+    private CancellationTokenRegistration? _applicationStartedRegistration;
+    private int _successfulCachedTokenCount;
+
+    public CacheTokenOnStartupBackgroundService(
+        IHostApplicationLifetime applicationLifetime,
+        IClientCredentialsTokenManagementService tokenManagementService,
+        IOptions<CacheTokenOnStartupBackgroundServiceOptions> backgroundServiceOptions,
+        IOptionsMonitor<ClientCredentialsOptions> clientCredentialsOptionsMonitor)
+    {
+        this._applicationLifetime = applicationLifetime;
+        this._tokenManagementService = tokenManagementService;
+        this._backgroundServiceOptions = backgroundServiceOptions;
+        this._clientCredentialsOptionsMonitor = clientCredentialsOptionsMonitor;
+        this._allTokensCachedSignal = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+        this._clientNames = new List<string>();
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        this._clientNames.AddRange(this._backgroundServiceOptions.Value.ClientCredentialPoweredClientNames);
+
+        this._applicationStartedRegistration = this._applicationLifetime.ApplicationStarted.Register(() =>
+        {
+            foreach (var clientName in this._clientNames)
+            {
+                // We don't await this task because we want to cache all tokens in parallel
+                this.CacheTokenAsync(clientName, stoppingToken).Forget();
+            }
+        });
+
+        return Task.CompletedTask;
+    }
+
+    private async Task CacheTokenAsync(string clientName, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var options = this._clientCredentialsOptionsMonitor.Get(clientName);
+            if (!options.EnablePeriodicTokenBackgroundRefresh)
+            {
+                return;
+            }
+
+            _ = await this._tokenManagementService.GetAccessTokenAsync(clientName, CachingBehavior.ForceRefresh, cancellationToken).ConfigureAwait(false);
+
+            if (Interlocked.Increment(ref this._successfulCachedTokenCount) == this._clientNames.Count)
```

Comment:
Given the `ClientsNames` list is populated directly from the `ClientCredentialPoweredClientNames` property of the 
`CacheTokenOnStartupBackgroundServiceOptions` options and in the registrations in `AddClientCredentialsHandler` directly add a registered httpClient to the list.

Isn't there a possibility that `this._allTokensCachedSignal.SetResult(true);` will never be called if one of the clients hasn't enabled `EnablePeriodicTokenBackgroundRefresh` cause then I feel the `_successfulCachedTokenCount` will never equal `_clientNames.Count` no?

Reply:
Yep you're right, I'll fix it. I added EnablePeriodicTokenBackgroundRefresh at the end and didn't consider that point.